### PR TITLE
Add integration tests for device CRD validation

### PR DESCRIPTION
**Pull request by rohitsardesai83**
**What type of PR is this?**
/kind test

**What this PR does / why we need it**:
Add integration tests for validation of device crds.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

### CODE REVIEW

File: cloud/Makefile [hunk 1]
```diff
@@ -7,9 +7,20 @@ lint:
 	golangci-lint run --disable-all -E gofmt -E golint ./...
 	go vet ./...
 
+# Build and run tests.
+.PHONY: test
+ifeq ($(WHAT),)
+       TEST_DIR="./pkg/"
```
Comment (subpathdev) [inline cloud/Makefile]: I would use environment variables instead of the WHAT variable.
The execution of make could be:
make test WHAT=pkg (in your case)
TEST_DIR=pkg make test (in my case)
 
In the makefile this will be written as:
TEST_DIR ?= "./pkg/"

**Comment by subpathdev** [comment]
please add this test case to ci.

File: cloud/Makefile (hunk 1 above)
Comment (rohitsardesai83) [reply cloud/Makefile]: @subpathdev , I have changed the test scripts . Now there is no requirement of WHAT variable.
# Running unit tests
```sh
$make test
$// verbose mode
$make test GOFLAGS=-v
$// select a package to run tests
$make test WHAT=pkg/controller/manager/ GOFLAGS=-v
$// select a test case to run
$make test GOFLAGS="-v" KUBE_TEST_ARGS="-run ^TestEventToMessage"
```
# Running integration tests
```sh
$make integration-test
$//verbose mode
$make integration-test GOFLAGS=-v
$// select a package to run tests
$make test-integration GOFLAGS=-v KUBE_TEST_ARGS="-run ^TestInvalidDeviceModel$"
$// select a folder to run tests
$make test-integration GOFLAGS=-v WHAT=./test/integration/crds/
```

**Comment by rohitsardesai83** [comment]
> please add this test case to ci.

Currently , integration tests require a Kubernetes master to be up and running for applying the crds, running tests , and cleanup.
The CI environment runs all e2e tests against a common Kubernetes master running on a vm. 
We are working on making each e2e test run in a separate container . This way parallel builds can run without affecting common  resources on which they operate.
One this in effect , we can also enable integration tests to run in a similar fashion.
For now , I suggest to get this PR merged and address the CI inclusion as a separate PR.

**Comment by rohitsardesai83** [comment]
@pavan187 , @sujithsimon22 , we need to include these integration tests in the CI once @sujithsimon22 finishes the task of running the e2e tests in separate container on each run and avoiding the dependency on a separate common K8S master.

**Comment by rohitsardesai83** [comment]
please review @kevin-wangzefeng

**Comment by subpathdev** [comment]
> Currently , integration tests require a Kubernetes master to be up and running for applying the crds, running tests , and cleanup.
> The CI environment runs all e2e tests against a common Kubernetes master running on a vm.
> We are working on making each e2e test run in a separate container . This way parallel builds can run without affecting common resources on which they operate.
> One this in effect , we can also enable integration tests to run in a similar fashion.
> For now , I suggest to get this PR merged and address the CI inclusion as a separate PR.

OK
\lgtm

File: cloud/hack/test-integration.sh [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  done
+
+  logStatus "Cleanup complete"
+}
+
+logStatus() {
+  timestamp=$(date +"[%m%d %H:%M:%S]")
+  echo "+++ ${timestamp} ${1}"
+  shift
+  for message; do
```
Comment (fisherxu) [inline cloud/hack/test-integration.sh]: what does the `message` mean here?

File: cloud/hack/test-integration.sh [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+      WHAT="${WHAT:-$(findIntegrationTestDirs | paste -sd' ' -)}" \
+      GOFLAGS="${GOFLAGS:-}" \
+      KUBE_TEST_ARGS="--alsologtostderr=true ${KUBE_TEST_ARGS:-} " \
+      KUBE_RACE="" \
+
+  cleanup
+}
+
+findIntegrationTestDirs() {
+
```
Comment (fisherxu) [inline cloud/hack/test-integration.sh]: could we delete these empty lines in the funcs?

File: cloud/hack/test-integration.sh (hunk 2 above)
Comment (rohitsardesai83) [reply cloud/hack/test-integration.sh]: this is not needed. removed.

File: cloud/hack/test-integration.sh (hunk 3 above)
Comment (rohitsardesai83) [reply cloud/hack/test-integration.sh]: done. thanks.

File: cloud/hack/test.sh [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+  )
+}
+
+
+testcases=()
+for arg; do
+  if [[ "${arg}" == -* ]]; then
+    goflags+=("${arg}")
+  else
+    testcases+=("${arg}")
```
Comment (fisherxu) [inline cloud/hack/test.sh]: where's the testcases used when runTests

File: cloud/hack/test-integration.sh [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+runTests() {
+  cleanup
+
+  logStatus "Applying device crd yamls"
+  kubectl create -f "${SCRIPT_ROOT}/build/crds/devices/devices_v1alpha1_devicemodel.yaml"
+  kubectl create -f "${SCRIPT_ROOT}/build/crds/devices/devices_v1alpha1_device.yaml"
+
+  logStatus  "Running integration test cases"
+  export KUBE_RACE="-race"
```
[the remark this answers is not in the document]
Comment (rohitsardesai83) [reply cloud/hack/test-integration.sh]: it is passed on line 38 to enable the Go race detector while running tests.

File: cloud/hack/test.sh [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+set -o errexit
+set -o nounset
+set -o pipefail
+
+KUBE_RACE=${KUBE_RACE:-}
```
[the remark this answers is not in the document]
Comment (rohitsardesai83) [reply cloud/hack/test.sh]: this argument is passed from the integration test script when it calls the make test target.

File: cloud/hack/test-integration.sh [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+    cd "${SCRIPT_ROOT}/cloud"
+    find test/integration/ -name '*_test.go' -print0 \
+      | xargs -0n1 dirname | LC_ALL=C sort -u
+  )
+}
+
+cleanup() {
+  logStatus "Cleaning up"
+
+  local num_devices=$(kubectl get device | wc -l) && rc=$?
```
Comment (kevin-wangzefeng) [inline cloud/hack/test-integration.sh]: Maybe use a random namespace for testing and tear it down afterwards?
Manually deleting the devices and devicemodels look inflexible.

File: cloud/hack/test.sh (hunk 4 above)
Comment (rohitsardesai83) [reply cloud/hack/test.sh]: testcases are used to run unit tests.  When the integration tests are run , these are not useful.
however , when unit tests need to be run , the correct test directories are found and set in testcases variable.  this variable is then set as an argument on line 49.

File: cloud/hack/test-integration.sh (hunk 7 above)
Comment (rohitsardesai83) [reply cloud/hack/test-integration.sh]: Yes, ideally we should be doing this programmatically, however, when I tried using the 
k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset libraries, I ran into dependency issues which can't be resolved until we upgrade to new Kubernetes version.

So I have kept the setup and teardown as part of the scripts itself.

File: cloud/hack/test-integration.sh [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+  kubectl create -f "${SCRIPT_ROOT}/build/crds/devices/devices_v1alpha1_devicemodel.yaml"
+  kubectl create -f "${SCRIPT_ROOT}/build/crds/devices/devices_v1alpha1_device.yaml"
+
+  logStatus  "Running integration test cases"
+  export KUBE_RACE="-race"
+  make -C "${SCRIPT_ROOT}/cloud" test \
+      WHAT="${WHAT:-$(findIntegrationTestDirs | paste -sd' ' -)}" \
+      GOFLAGS="${GOFLAGS:-}" \
+      KUBE_TEST_ARGS="--alsologtostderr=true ${KUBE_TEST_ARGS:-} " \
+      KUBE_RACE="" \
```
[the remark this answers is not in the document]
Comment (rohitsardesai83) [reply cloud/hack/test-integration.sh]: fixed. thanks.

File: cloud/hack/test-integration.sh (hunk 7 above)
Comment (fisherxu) [reply cloud/hack/test-integration.sh]: we just verify the validation of device CRD here, how about run and set --dryrun flag, so no need do the cleanup

Comment (rohitsardesai83) [reply cloud/hack/test-integration.sh]: The CRD definitions need to be installed in the cluster in order to create the device model and instances and validate.

**Comment by rohitsardesai83** [comment]
all review comments have been address. @fisherxu , @subpathdev , @sids-b , @chendave PTAL.

File: cloud/hack/test-integration.sh [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+      | xargs -0n1 dirname | LC_ALL=C sort -u
+  )
+}
+
+cleanup() {
+  logStatus "Cleaning up"
+
+  local num_devices=$(kubectl get device | wc -l) && rc=$?
+  if [[ num_devices -gt 0 ]]; then
+    declare -a devices=("device-modbus-rtu" "device-modbus-tcp" "device-opcua" )
```
Comment (chendave) [inline cloud/hack/test-integration.sh]: Can we get the device directly from "kubectl get device"? so that the loop here might not be needed.

File: cloud/hack/test-integration.sh [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    declare -a device_models=("bluetooth-device-model" "modbus-device-model" "opcua-device-model" )
+    for device_model in "${device_models[@]}"; do
+      kubectl delete devicemodel ${device_model}
+    done
+  fi
+
+  declare -a device_crds=("devicemodels.devices.kubeedge.io" "devices.devices.kubeedge.io")
+
+  for device_crd in "${device_crds[@]}"; do
+    crd_num=$(kubectl get crd ${device_crd} | wc -l) && rc=$?
```
Comment (chendave) [inline cloud/hack/test-integration.sh]: might be like this?

File: cloud/test/integration/crds/validation_test.go [hunk 11]
```diff
@@ -0,0 +1,342 @@
...
+		},
+		"valid opc ua device model": {
+			deviceModelFn: func() *v1alpha1.DeviceModel {
+				deviceModel := fixtures.NewDeviceModelOpcUA("opcua-device-model", fixtures.DefaultNamespace)
+				return deviceModel
+			},
+		},
+	}
+
+	crdClient := buildCrdClient(t)
```
Comment (chendave) [inline cloud/test/integration/crds/validation_test.go]: I guess you want to ensure the crdClient is not `nil`, or else L82 will see some exception when the `buildCrdClient` return `nil`.

File: cloud/test/integration/crds/validation_test.go [hunk 12]
```diff
@@ -0,0 +1,342 @@
...
+		},
+		"device model with opc ua property no nodeID": {
+			deviceModelFn: func() *v1alpha1.DeviceModel {
+				deviceModel := fixtures.NewDeviceModelOpcUANoNodeID("model-modbus-no-nodeID", fixtures.DefaultNamespace)
+				return deviceModel
+			},
+		},
+	}
+
+	crdClient := buildCrdClient(t)
```
Comment (chendave) [inline cloud/test/integration/crds/validation_test.go]: same here.

File: cloud/test/integration/crds/validation_test.go [hunk 13]
```diff
@@ -0,0 +1,342 @@
...
+				t.Fatalf("%s: expected nil err , got %v", name, result.Error())
+			}
+		})
+	}
+}
+
+func TestInvalidDeviceModel(t *testing.T) {
+
+	tests := map[string]struct {
+		deviceModelFn func() *v1alpha1.DeviceModel
```
Comment (chendave) [inline cloud/test/integration/crds/validation_test.go]: could this be consistent with `TestValidDeviceModel` where "name" is defined in the struct?

File: cloud/hack/test-integration.sh (hunk 9 above)
Comment (rohitsardesai83) [reply cloud/hack/test-integration.sh]: the loop is for avoiding 3 separate lines for deleting every device. Without loop, we need to do the below : 
e.g
```sh
kubectl delete device device-modbus-rtu
kubectl delete device device-modbus-tcp
kubectl delete device device-opcua
```

File: cloud/test/integration/crds/validation_test.go (hunk 11 above)
Comment (rohitsardesai83) [reply cloud/test/integration/crds/validation_test.go]: in the buildCrdClient() method , if the client construction fails , t.Fatalf will generate an exception and the test won't be executed further. So line 82 won't be hit.

```sh
--- FAIL: TestValidDeviceModel (0.00s)
    validation_test.go:37: Failed to build device CRD client, error is invalid configuration: no configuration has been provided
```

File: cloud/test/integration/crds/validation_test.go (hunk 13 above)
Comment (rohitsardesai83) [reply cloud/test/integration/crds/validation_test.go]: actually the name field was not used . removed it from TestValidDeviceModel test.

File: cloud/hack/test-integration.sh (hunk 9 above)
Comment (chendave) [reply cloud/hack/test-integration.sh]: What I meant was there might be only one device, but it always loops three times to delete the potential not existing device, for example, `kubectl get device | awk -F' ' '{print $1}' | tail -2` will get all the devices that is actually created and then delete them.

File: cloud/test/integration/crds/validation_test.go (hunk 13 above)
Comment (chendave) [reply cloud/test/integration/crds/validation_test.go]: +1, pls remove it if it's not used.

File: cloud/test/integration/crds/validation_test.go (hunk 11 above)
Comment (chendave) [reply cloud/test/integration/crds/validation_test.go]: yes, I was misleading by line 38 where it return `nil` on error.

File: cloud/test/integration/fixtures/devicemodel.go [hunk 14]
```diff
@@ -0,0 +1,401 @@
...
+func DeviceModelWithPropertyNoType(name string, namespace string) *v1alpha1.DeviceModel {
+
+	deviceModel := newDeviceModel(name, namespace)
+	devicePropertyOp := newDevicePropertyOp(withName(devicePropertyTemperature), withDescription(devicePropertyTemperatureDesc))
+	deviceModel.Spec.Properties = append(deviceModel.Spec.Properties, devicePropertyOp.deviceProperty)
+	return deviceModel
+}
+
+func DeviceModelWithPropertyBadAccessMode(name string, namespace string) *v1alpha1.DeviceModel {
+
```
Comment (chendave) [inline cloud/test/integration/fixtures/devicemodel.go]: could you pls remove those white spaces?

File: cloud/test/integration/fixtures/device.go [hunk 15]
```diff
@@ -0,0 +1,289 @@
...
+}
+
+func withOPCUAServerURL(url string) DeviceOption {
+	return func(op *DeviceOp) {
+		op.device.Spec.Protocol.OpcUA.URL = url
+	}
+}
+
+func NewDeviceModbusRTU(name string, namespace string) v1alpha1.Device {
+
```
Comment (chendave) [inline cloud/test/integration/fixtures/device.go]: pls remove those white-spaces.

File: cloud/pkg/cloudhub/common/model/types_test.go [hunk 16]
```diff
@@ -84,8 +85,9 @@ func TestEventToMessage(t *testing.T) {
 
 // TestMessageToEvent is function to test MessageToEvent
 func TestMessageToEvent(t *testing.T) {
-	msg := modelMessage("ID1", "PID1", time.Now().UnixNano()/1e6, "Source1", "Group1", "Operation1", "Resource1")
-	event := modelEvent("ID1", "PID1", time.Now().UnixNano()/1e6, "Source1", "Group1", "Operation1", "Resource1", nil)
+	timestamp := time.Now().UnixNano() / 1e6
+	msg := modelMessage("ID1", "PID1", timestamp, "Source1", "Group1", "Operation1", "Resource1")
+	event := modelEvent("ID1", "PID1", timestamp, "Source1", "Group1", "Operation1", "Resource1", nil)
```
Comment (chendave) [inline cloud/pkg/cloudhub/common/model/types_test.go]: Those refactoring might be better in another PR since it is irrelevant with this topic, good as it as.

File: cloud/hack/test-integration.sh [hunk 17]
```diff
@@ -0,0 +1,87 @@
...
+set -o errexit
+set -o nounset
+set -o pipefail
+
+SCRIPT_ROOT=$(unset CDPATH && cd $(dirname "${BASH_SOURCE[0]}")/../../ && pwd)
+export KUBE_CONFIG=""
+export KUBE_APISERVER_URL="http://localhost:8080"
+
+runTests() {
+  cleanup
```
Comment (chendave) [inline cloud/hack/test-integration.sh]: all resource should be created in a test namespace, and cleanup should be done in the test namespace as well.

File: cloud/hack/test-integration.sh (hunk 9 above)
Comment (rohitsardesai83) [reply cloud/hack/test-integration.sh]: thanks @chendave for the inputs , I have made the suggested changes. Thanks.

File: cloud/test/integration/fixtures/devicemodel.go (hunk 14 above)
Comment (rohitsardesai83) [reply cloud/test/integration/fixtures/devicemodel.go]: done

File: cloud/pkg/cloudhub/common/model/types_test.go (hunk 16 above)
Comment (rohitsardesai83) [reply cloud/pkg/cloudhub/common/model/types_test.go]: the unit tests were failing when i ran the test via scripts. so I fixed them here.

File: cloud/hack/test-integration.sh (hunk 17 above)
Comment (rohitsardesai83) [reply cloud/hack/test-integration.sh]: created a test namespace. thanks

File: cloud/hack/test-integration.sh [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+
+SCRIPT_ROOT=$(unset CDPATH && cd $(dirname "${BASH_SOURCE[0]}")/../../ && pwd)
+export KUBE_CONFIG=""
+export KUBE_APISERVER_URL="http://localhost:8080"
+export TESTNS="crd-test"
+
+runTests() {
+  logStatus "Applying device crd yamls"
+  kubectl create -f "${SCRIPT_ROOT}/build/crds/devices/devices_v1alpha1_devicemodel.yaml"
+  kubectl create -f "${SCRIPT_ROOT}/build/crds/devices/devices_v1alpha1_device.yaml"
```
Comment (chendave) [inline cloud/hack/test-integration.sh]: Could you pls isolate those resource in the namespace as well?

```bash
 make test-integration
hack/test-integration.sh
+++ [0716 10:04:16] Applying device crd yamls
Error from server (AlreadyExists): error when creating "/go/src/github.com/kubeedge/kubeedge/build/crds/devices/devices_v1alpha1_devicemodel.yaml": customresourcedefinitions.apiextensions.k8s.io "devicemodels.devices.kubeedge.io" already exists
Makefile:40: recipe for target 'test-integration' failed
make: *** [test-integration] Error 1
```

File: cloud/pkg/cloudhub/common/model/types_test.go (hunk 16 above)
Comment (chendave) [reply cloud/pkg/cloudhub/common/model/types_test.go]: It's fine.

File: cloud/hack/test-integration.sh (hunk 18 above)
Comment (rohitsardesai83) [reply cloud/hack/test-integration.sh]: the crd definitions are cluster wide resourcesand not tied to a particular namespace.

**Comment by rohitsardesai83** [comment]
> pls check my comments inline, or let me know if I miss something, thanks.

thanks @chendave , addressed your comments.

File: cloud/hack/test-integration.sh (hunk 18 above)
Comment (chendave) [reply cloud/hack/test-integration.sh]: How about some pre-check before the creation?

Comment (rohitsardesai83) [reply cloud/hack/test-integration.sh]: added the check. thanks.

**Comment by chendave** [comment]
@rohitsardesai83 , thanks for addressing my comments, verified locally and this PR works very well!

**Comment by rohitsardesai83** [comment]
ping @kevin-wangzefeng , @fisherxu , @sids-b for review.
